File: cpp/67-AddBinary/main.cpp
#include <string>
#include <iostream>
#include <algorithm>

using namespace std;

class Solution 
{
private:
    int nextDigit(const string &s, int index) const
    {
        if(index >= 0) return s[index] - '0';

        return 0;
    }

public:
    string addBinary(string a, string b) 
    {
        string result = "";
        
        int offsetA = a.size();
        --offsetA;

        int offsetB = b.size();
        --offsetB;
        
        int carry = 0;

        while(offsetA >= 0 || offsetB >= 0 || carry == 1)
        {
            const auto bitA = nextDigit(a, offsetA);
            const auto bitB = nextDigit(b, offsetB);
            
            const auto sum = bitA + bitB + carry;
            result.push_back('0' + (sum % 2));            
            carry = sum / 2;

            offsetA--;
            offsetB--;
        }

        reverse(result.begin(), result.end());
        return result;
    }
};

int main()
{
    Solution s;
    std::cout << s.addBinary("11", "1") << "\n";
    std::cout << s.addBinary("1010", "1011") << "\n";
}

File: go/67-AddBinary/main.go
package main

import (
	"fmt"
	"strings"
)

func nextDigit(s string, index int) byte {
	if index >= 0 {
		return s[index] - byte(int('0'))
	}

	return 0
}

func addBinary(a string, b string) string {
	buffer := make([]byte, 0, len(a)+len(b)+1)

	// We need to add the characters from right to left
	offsetA := len(a) - 1
	offsetB := len(b) - 1
	carry := byte(0)

	for offsetA >= 0 || offsetB >= 0 || carry == 1 {
		bitA := nextDigit(a, offsetA)
		bitB := nextDigit(b, offsetB)

		sum := bitA + bitB + carry
		buffer = append(buffer, sum%2)
		carry = sum / 2

		offsetA--
		offsetB--
	}

	var builder strings.Builder
	builder.Grow(len(buffer))

	// The bits are back to front
	for i := len(buffer) - 1; i >= 0; i-- {
		builder.WriteByte(byte('0') + buffer[i])
	}

	return builder.String()
}

func main() {
	fmt.Println(addBinary("11", "1"))
	fmt.Println(addBinary("1010", "1011"))
}
